Build failure: pbzip2

## Steps To Reproduce

Steps to reproduce the behavior:

1. build pbzip2

## Build log

<details>

<summary>Build Log</summary>

```
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/s6nrggkvaw6mxlwxdw5zn9avz796vnm9-pbzip2-1.1.13.tar.gz
source root is pbzip2-1.1.13
setting SOURCE_DATE_EPOCH to timestamp 1450395169 of file pbzip2-1.1.13/releasing/set_version.rb
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
no configure script, doing nothing
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
substituteStream() in derivation pbzip2-1.1.13: WARNING: '--replace' is deprecated, use --replace-{fail,warn,quiet}. (file 'Makefile')
build flags: SHELL=/nix/store/5wjj1f3pa1ndi3kvskm1fcs9kdy2xajf-bash-5.2p37/bin/bash
clang++ -O2 -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -DUSE_STACKSIZE_CUSTOMIZATION -pthread -D_POSIX_PTHREAD_SEMANTICS  pbzip2.cpp BZ2StreamScanner.cpp ErrorContext.cpp -o pbzip2 -lbz2 -lpthread
In file included from pbzip2.cpp:13:
In file included from ./pbzip2.h:16:
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:820:42: error: implicit instantiation of undefined template 'std::char_traits<unsigned char>'
  820 |   static_assert(is_same<_CharT, typename traits_type::char_type>::value,
      |                                          ^
./BZ2StreamScanner.h:128:25: note: in instantiation of template class 'std::basic_string<unsigned char>' requested here
  128 |         basic_string<CharType> _bz2Header;
      |                                ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
In file included from pbzip2.cpp:13:
In file included from ./pbzip2.h:16:
In file included from /nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:647:
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string_view:300:42: error: implicit instantiation of undefined template 'std::char_traits<unsigned char>'
  300 |   static_assert(is_same<_CharT, typename traits_type::char_type>::value,
      |                                          ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__type_traits/is_convertible.h:22:99: note: in instantiation of template class 'std::basic_string_view<unsigned char>' requested here
   22 | struct _LIBCPP_TEMPLATE_VIS is_convertible : public integral_constant<bool, __is_convertible(_T1, _T2)> {};
      |                                                                                                   ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:745:29: note: in instantiation of template class 'std::is_convertible<const std::basic_string<unsigned char> &, std::basic_string_view<unsigned char>>' requested here
  745 |     : public _BoolConstant< is_convertible<const _Tp&, basic_string_view<_CharT, _Traits> >::value &&
      |                             ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:1223:27: note: in instantiation of template class 'std::__can_be_converted_to_string_view<unsigned char, std::char_traits<unsigned char>, std::basic_string<unsigned char>>' requested here
 1223 |             __enable_if_t<__can_be_converted_to_string_view<_CharT, _Traits, _Tp>::value &&
      |                           ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:1226:47: note: while substituting prior template arguments into non-type template parameter [with _Tp = std::basic_string<unsigned char>]
 1226 |   _LIBCPP_CONSTEXPR_SINCE_CXX20 basic_string& operator=(const _Tp& __t) {
      |                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~
 1227 |     __self_view __sv = __t;
      |     ~~~~~~~~~~~~~~~~~~~~~~~
 1228 |     return assign(__sv);
      |     ~~~~~~~~~~~~~~~~~~~~
 1229 |   }
      |   ~
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:752:7: note: while substituting deduced template arguments into function template 'operator=' [with _Tp = std::basic_string<unsigned char>, $1 = (no value)]
  752 | class basic_string {
      |       ^
./BZ2StreamScanner.h:20:7: note: while declaring the implicit copy assignment operator for 'BZ2StreamScanner'
   20 | class BZ2StreamScanner
      |       ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
pbzip2.cpp:1223:75: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
 1223 |                                 fprintf(stderr, "pbzip2:  *WARNING: Compressed block size is large [%"PRIuMAX" bytes].\n",
      |                                                                                                       ^
      |                                                                                                        
pbzip2.cpp:1832:39: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
 1832 |                 fprintf(stderr, "    Output Size: %"PRIuMAX" bytes\n", (uintmax_t)CompressedSize);
      |                                                     ^
      |                                                      
pbzip2.cpp:2075:39: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
 2075 |                 fprintf(stderr, "    Output Size: %"PRIuMAX" bytes\n", (uintmax_t)CompressedSize);
      |                                                     ^
      |                                                      
pbzip2.cpp:2797:80: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
 2797 |                 fprintf(stderr, "pbzip2: *ERROR: Could not initialize (OutputBuffer); size=%"PRIuMAX"!  Aborting...\n", (uintmax_t)size);
      |                                                                                              ^
      |                                                                                               
pbzip2.cpp:4105:41: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
 4105 |                                 fprintf(stderr, "     Input Size: %"PRIuMAX" bytes\n", (uintmax_t)InFileSize);
      |                                                                     ^
      |                                                                      
pbzip2.cpp:4227:65: warning: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
 4227 |                                         fprintf(stderr, "*Warning* Max memory limit increased to %"PRIuMAX" MB to support %d CPUs\n", (uintmax_t)((NumBufferedBlocksMax + (numCPU * 2)) * blockSize)/1000000, numCPU);
      |                                                                                                    ^
      |                                                                                                     
6 warnings and 2 errors generated.
In file included from BZ2StreamScanner.cpp:8:
In file included from ./pbzip2.h:16:
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:820:42: error: implicit instantiation of undefined template 'std::char_traits<unsigned char>'
  820 |   static_assert(is_same<_CharT, typename traits_type::char_type>::value,
      |                                          ^
./BZ2StreamScanner.h:128:25: note: in instantiation of template class 'std::basic_string<unsigned char>' requested here
  128 |         basic_string<CharType> _bz2Header;
      |                                ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
In file included from BZ2StreamScanner.cpp:8:
In file included from ./pbzip2.h:16:
In file included from /nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:647:
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string_view:300:42: error: implicit instantiation of undefined template 'std::char_traits<unsigned char>'
  300 |   static_assert(is_same<_CharT, typename traits_type::char_type>::value,
      |                                          ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__type_traits/is_convertible.h:22:99: note: in instantiation of template class 'std::basic_string_view<unsigned char>' requested here
   22 | struct _LIBCPP_TEMPLATE_VIS is_convertible : public integral_constant<bool, __is_convertible(_T1, _T2)> {};
      |                                                                                                   ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:745:29: note: in instantiation of template class 'std::is_convertible<const std::basic_string<unsigned char> &, std::basic_string_view<unsigned char>>' requested here
  745 |     : public _BoolConstant< is_convertible<const _Tp&, basic_string_view<_CharT, _Traits> >::value &&
      |                             ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:1223:27: note: in instantiation of template class 'std::__can_be_converted_to_string_view<unsigned char, std::char_traits<unsigned char>, std::basic_string<unsigned char>>' requested here
 1223 |             __enable_if_t<__can_be_converted_to_string_view<_CharT, _Traits, _Tp>::value &&
      |                           ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:1226:47: note: while substituting prior template arguments into non-type template parameter [with _Tp = std::basic_string<unsigned char>]
 1226 |   _LIBCPP_CONSTEXPR_SINCE_CXX20 basic_string& operator=(const _Tp& __t) {
      |                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~
 1227 |     __self_view __sv = __t;
      |     ~~~~~~~~~~~~~~~~~~~~~~~
 1228 |     return assign(__sv);
      |     ~~~~~~~~~~~~~~~~~~~~
 1229 |   }
      |   ~
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:752:7: note: while substituting deduced template arguments into function template 'operator=' [with _Tp = std::basic_string<unsigned char>, $1 = (no value)]
  752 | class basic_string {
      |       ^
./BZ2StreamScanner.h:20:7: note: while declaring the implicit copy assignment operator for 'BZ2StreamScanner'
   20 | class BZ2StreamScanner
      |       ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
In file included from BZ2StreamScanner.cpp:8:
In file included from ./pbzip2.h:16:
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:2825:29: error: implicit instantiation of undefined template 'std::char_traits<unsigned char>'
 2825 |            ? (__fits_in_sso(traits_type::length(__s)) ? __assign_short(__s, traits_type::length(__s))
      |                             ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:1243:12: note: in instantiation of member function 'std::basic_string<unsigned char>::assign' requested here
 1243 |     return assign(__s);
      |            ^
BZ2StreamScanner.cpp:52:13: note: in instantiation of member function 'std::basic_string<unsigned char>::operator=' requested here
   52 |         _bz2Header = bz2header;
      |                    ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
In file included from BZ2StreamScanner.cpp:8:
In file included from ./pbzip2.h:16:
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:2825:77: error: implicit instantiation of undefined template 'std::char_traits<unsigned char>'
 2825 |            ? (__fits_in_sso(traits_type::length(__s)) ? __assign_short(__s, traits_type::length(__s))
      |                                                                             ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
In file included from BZ2StreamScanner.cpp:8:
In file included from ./pbzip2.h:16:
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:2826:80: error: implicit instantiation of undefined template 'std::char_traits<unsigned char>'
 2826 |                                                       : __assign_external(__s, traits_type::length(__s)))
      |                                                                                ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
In file included from BZ2StreamScanner.cpp:8:
In file included from ./pbzip2.h:16:
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:2208:5: error: implicit instantiation of undefined template 'std::char_traits<unsigned char>'
 2208 |     traits_type::move(std::__to_address(__p), __s, __n);
      |     ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:2825:57: note: in instantiation of member function 'std::basic_string<unsigned char>::__assign_short' requested here
 2825 |            ? (__fits_in_sso(traits_type::length(__s)) ? __assign_short(__s, traits_type::length(__s))
      |                                                         ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:1243:12: note: in instantiation of member function 'std::basic_string<unsigned char>::assign' requested here
 1243 |     return assign(__s);
      |            ^
BZ2StreamScanner.cpp:52:13: note: in instantiation of member function 'std::basic_string<unsigned char>::operator=' requested here
   52 |         _bz2Header = bz2header;
      |                    ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
In file included from BZ2StreamScanner.cpp:8:
In file included from ./pbzip2.h:16:
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:2209:5: error: implicit instantiation of undefined template 'std::char_traits<unsigned char>'
 2209 |     traits_type::assign(__p[__n], value_type());
      |     ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
In file included from BZ2StreamScanner.cpp:8:
In file included from ./pbzip2.h:16:
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:2817:33: error: implicit instantiation of undefined template 'std::char_traits<unsigned char>'
 2817 |   return __assign_external(__s, traits_type::length(__s));
      |                                 ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:2827:14: note: in instantiation of member function 'std::basic_string<unsigned char>::__assign_external' requested here
 2827 |            : __assign_external(__s);
      |              ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:1243:12: note: in instantiation of member function 'std::basic_string<unsigned char>::assign' requested here
 1243 |     return assign(__s);
      |            ^
BZ2StreamScanner.cpp:52:13: note: in instantiation of member function 'std::basic_string<unsigned char>::operator=' requested here
   52 |         _bz2Header = bz2header;
      |                    ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
In file included from BZ2StreamScanner.cpp:8:
In file included from ./pbzip2.h:16:
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/string:3736:22: error: implicit instantiation of undefined template 'std::char_traits<unsigned char>'
 3736 |   int __r          = traits_type::compare(data() + __pos1, __s, std::min(__rlen, __n2));
      |                      ^
BZ2StreamScanner.cpp:364:24: note: in instantiation of member function 'std::basic_string<unsigned char>::compare' requested here
  364 |                 else if ( _bz2Header.compare( 0, prefixLen, getInBuffSearchPtr(), prefixLen ) == 0 )
      |                                      ^
/nix/store/0nhi47d5ip48wprxnava6vv973zzzndf-libcxx-19.1.5-dev/include/c++/v1/__fwd/string.h:23:29: note: template is declared here
   23 | struct _LIBCPP_TEMPLATE_VIS char_traits;
      |                             ^
9 errors generated.
make: *** [Makefile:70: pbzip2] Error 1
```

</details>

## Metadata

 - system: `"aarch64-darwin"`
 - host os: `Darwin 24.1.0, macOS 15.1.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Lix, like Nix) 2.91.1
System type: aarch64-darwin
Additional system types: x86_64-darwin
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /Users/wcampbel/.config/nix/nix.conf:/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/4cz2w0kc0vgnvwm2643dnh0rmhxs0yr0-lix-2.91.1/share`
 - channels(root): `""`
 - nixpkgs: `/Users/wcampbel/.local/state/nix/profiles/channels/nixpkgs`

The channel is `nixpkgs-unstable`

## Notify maintainers

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/pb/pbzip2/package.nix
+++ b/pkgs/by-name/pb/pbzip2/package.nix
@@ -2,6 +2,7 @@
   lib,
   stdenv,
   fetchurl,
+  fetchpatch,
   bzip2,
 }:
 
@@ -18,13 +19,29 @@ stdenv.mkDerivation rec {
     sha256 = "1rnvgcdixjzbrmcr1nv9b6ccrjfrhryaj7jwz28yxxv6lam3xlcg";
   };
 
+  patches = [
+    # https://libcxx.llvm.org/ReleaseNotes/19.html#deprecations-and-removals
+    # https://bugs.launchpad.net/pbzip2/+bug/2081588
+    (fetchpatch {
+      url = "https://github.com/freebsd/freebsd-ports/raw/974d3ff054965d2bd2ab884a0579ed06c5a08b07/archivers/pbzip2/files/patch-BZ2StreamScanner.cpp";
+      extraPrefix = "";
+      hash = "sha256-dvXdp+5S41akavy+mvPGHpUxHxenXS7bbTVBVkIJj0s=";
+    })
+    (fetchpatch {
+      url = "https://github.com/freebsd/freebsd-ports/raw/974d3ff054965d2bd2ab884a0579ed06c5a08b07/archivers/pbzip2/files/patch-BZ2StreamScanner.h";
+      extraPrefix = "";
+      hash = "sha256-/twP8HyHP4cAVgb5cUPq0CgDxUgDYPdd9haH9wDOrz8=";
+    })
+  ];
+
+  postPatch = ''
+    substituteInPlace pbzip2.cpp \
+      --replace-fail '"PRIuMAX"' '" PRIuMAX "'
+  '';
+
   buildInputs = [ bzip2 ];
 
-  preBuild = "substituteInPlace Makefile --replace g++ c++";
-
   installFlags = [ "PREFIX=$(out)" ];
-
-  env.NIX_CFLAGS_COMPILE = lib.optionalString stdenv.cc.isClang "-Wno-error=reserved-user-defined-literal";
 
   meta = with lib; {
     homepage = "http://compression.ca/pbzip2/";

